pkg/controller/volume/attachdetach/cache/actual_state_of_world.go

```diff
@@ -517,7 +519,7 @@ func (asw *actualStateOfWorld) addVolumeToReportAsAttached(
 // needs to be updated again by the node status updater.
 // If the specified node does not exist in the nodesToUpdateStatusFor list, log the error and return
 // This is an internal function and caller should acquire and release the lock
-func (asw *actualStateOfWorld) updateNodeStatusUpdateNeeded(nodeName types.NodeName, needed bool) error {
+func (asw *actualStateOfWorld) UpdateNodeStatusUpdateNeeded(nodeName types.NodeName, needed bool) error {
```

Comment:
@jingxu97 I am not very sure about what should be done between 
- make new files in statusupdater folder and include `updateNodeStatusUpdateNeeded` function in it or allow `updateNodeStatusUpdateNeeded` function to be exported in other packages by changing its name to `UpdateNodeStatusUpdateNeeded`.

I executed the 2nd approach, could you please give your thoughts.

Reply:
Let me drive-by with a couple of comments

The first is structural: you don't want to be exposing updateNodeStatusUpdateNeeded publically; note that when used in this package it's always protected by asw.Lock(). By just making this public, you've destroyed that guarantee which will cause big problems in any concurrent context.

The second is that I don't think your changes are fixing the main problem, which is that the nodes are set as updated before the update happens, meaning that if something goes wrong halfway through they are incorrectly marked as updated when in fact they still need it. If I understood Jing's point in the issue correctly, I think what we actually want to do is remove all the updateNodeStatusUpdateNeeded(.., false) calls in actual_state_of_world.go, and then in processNodeVolumes in node_status_updater.go, instead of calling SetNodeStatusUpdateNeeded(nodeName) if there is a failure, call (a new method) SetNodeStatusUpdateNotNeeded(nodeName) if the updateNodeStatusCall succeeds.

I'm not sure why it wasn't done this way in the first place as that seems more obvious. Is there some case where we want to stop retrying, or something like that? You may have to look through all the code paths and see if SetNodeStatusUpdatedNotNeeded should be called somewhere else (where effectively it's currently implicitly called, because we update the node statuses before we start processing the nodes).